### PR TITLE
Chado custom table functionality

### PR DESCRIPTION
**Pull request by risharde**
Tested and seems to be working when using the UI and also tested the chado create custom table api function via Devel.

### CODE REVIEW

**Comment by risharde** [comment]
I just realized that this code may need to be further adjusted now to use bio_db. Just want to confirm this before proceeding at some point on this. @laceysanderson @spficklin 
IIRC this was during the time we were using chado_query for api functions so even though it's cherry picked to be included directly to core, it uses chado_query OR the getConnection('default') which I think we want to upgrade?

@spficklin @laceysanderson Actually double checking this, I see the chado_custom api is using the bio_db implementation so this was probably a lapse in my memory when I adjusted it. I think this works as intended!

**Comment by spficklin** [comment]
Here's what I did to test this:

- Chado Custom Tables UI
  - [x] Adding a new custom table adds the table to Chado and the `tripal_custom_tables` table.
  - [x] Editing a custom table and only change the table name.  The table should be renamed in Chado and the record in the `tripal_custom_tables` table should be updated
  - [x] Changing a field in the schema without checking the 're-create' button should give a message that the schema definition was updated but the table was not.  The schema should be updated in the `tripal_custom_tables` table.
  - [x] Changing a field in the schema and checking the 're-create' button should drop the table, recreate it in Chado and update the schema in the `tripal_custom_tables` table.  
  - [x] Deleting a custom table should remove it from Chado and from the `tripal_custom_tables` table.
  - [x] Filtering by table name works

Things that need fixing:

- [x]  The menu item `Admin -> Tripal -> Data Storage -> Chado Custom Table` should go away.  The user will create a custom table via a link the "Chado Custom Tables" page.
- The Chado Custom Tables UI Page
  - [x]  When there are no custom tables page should still show the table but indicate there are no current custom tables.
  - [x] The "Add new chado custom table" should look like a blue button with a plus (See an example on the "Page Structure" page)
- The Add New Chado Custom Table page
  - [x] The "Re-create" table is not needed when creating a new custom table.  
  - [x] The example should have new PHP style arrays.
  - [x] When adding or editing a new table it should give a Drupal status message indicating that the table creation/update was successful.
- The Edit Chado Custom Table page
  - [x] When editing a table without checking the "Re-create table" checkbox nothing changes with the array that is saved and there is no message indicating anything was successful or failed.

There seems to be some missing parts (e.g. no Drupal messages) that were previously in the Tripal v3 code, so just double check the chado_create_custom_table

**Review by spficklin — CHANGES_REQUESTED** [review]
See the message in the conversation for requested changes.

**Comment by risharde** [comment]
The requirement for this one "Changing a field in the schema and checking the 're-create' button should drop the table, recreate it in Chado and update the schema in the tripal_custom_tables table." should work however in the test example, if one changes the primary key at the top of the array but forget to update the primary_key info at the bottom of the array - the table will not get created and an error message will be displayed that there was an error and the user should check the logs for further details on the error (which is showed in the SQL statement error)

**Comment by risharde** [comment]
Having problems doing this one with the blue button at the moment
"The "Add new chado custom table" should look like a blue button with a plus (See an example on the "Page Structure" page)"

**Comment by risharde** [comment]
For this task "The example should have new PHP style arrays.", I had to remove the check for the array string. Tested with the new php style arrays and it seems to work.

**Comment by risharde** [comment]
I finally got the action link to work! I learnt that action links don't seem to like to attach to view routes. Getting the route was also a challenge until I discovered you can easily find it using the Devel module.

**Comment by risharde** [comment]
@spficklin @laceysanderson Reminder to review please. I would like to get this in since the old PR which was intended to be pushed to the Importer Review branch contained bugs which are fixed in this one.

I also need this in to test the preparer task branch I'm working on which is probably going to be another PR that needs this updated version to work correctly.

Thanks!

**Review by spficklin — DISMISSED** [review]

**Comment by spficklin** [comment]
I approve of this PR.  All manual tests worked good. I did a code review and cleaned up a few items where older code was used instead of newer ones, removed unnecessary comments and updated the logging.

File: tripal_chado/src/Form/ChadoCustomTableForm.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    // set the breadcrumb
+    
+    $breadcrumb = [];
+    $breadcrumb[] = Link::fromTextAndUrl('Home', Url::fromRoute('<front>'));
+    $breadcrumb[] = Link::fromTextAndUrl('Administration', Url::fromUserInput('/admin'));
+    $breadcrumb[] = Link::fromTextAndUrl('Tripal', Url::fromUserInput('/admin/tripal'));
+    $breadcrumb[] = Link::fromTextAndUrl('Chado Schema', Url::fromUserInput('/admin/tripal/storage/chado'));
+    $breadcrumb[] = Link::fromTextAndUrl('Custom Tables', Url::fromUserInput('/admin/tripal/storage/chado/custom_tables'));
+    // TODO D9
+    // drupal_set_breadcrumb($breadcrumb);
```
Comment (laceysanderson) [inline tripal_chado/src/Form/ChadoCustomTableForm.php]: I noticed you set the breadcrumb in the controller. Does that mean this section of code can be removed?

File: tripal_chado/src/Form/ChadoCustomTableForm.php [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    $force_drop = $values['force_drop'];
+  
+    if (!$schema) {
+      $form_state->setErrorByName($values['schema'], t('Schema array field is required.'));
+    }
+  
+    // make sure the array is valid
+    $schema_array = [];
+    if ($schema) {
+      // $success = preg_match('/^\s*array/', $schema);
```
Comment (laceysanderson) [inline tripal_chado/src/Form/ChadoCustomTableForm.php]: This looks like you just removed the check for a valid array rather then re-implementing it?

File: tripal_chado/src/api/ChadoSchema.php [hunk 3]
```diff
@@ -247,6 +247,12 @@ public function getTableSchema($table) {
     }
     else {
       $table_arr =  FALSE;
+      // Try to check if it's a custom table
+      $table_arr = $this->getCustomTableSchema($table);
+      // print_r($table_arr);
+      if($table_arr == FALSE) {
+        $table_arr = FALSE; //TODO: Should this return false so below is not processed?
```
Comment (laceysanderson) [inline tripal_chado/src/api/ChadoSchema.php]: Come back to this one. Does it return false in the most recent dev version of T3?

File: tripal_chado/src/api/ChadoSchema.php [hunk 4]
```diff
@@ -278,7 +284,7 @@ public function getTableSchema($table) {
 
     // if the table_arr is empty then maybe this is a custom table
     if (!is_array($table_arr) or count($table_arr) == 0) {
-      //$table_arr = $this->getCustomTableSchema($table);
+      // $table_arr = $this->getCustomTableSchema($table);
```
Comment (laceysanderson) [inline tripal_chado/src/api/ChadoSchema.php]: Is there a reason why you are not fetching the table array here if it's not already set?

File: tripal_chado/src/api/tripal_chado.custom_tables.api.inc [hunk 5]
```diff
@@ -106,92 +137,96 @@ function chado_edit_custom_table($table_id, $table_name, $schema, $skip_if_exist
...
-                    
-    } catch (Exception $e) {
-        $transaction->rollback();
+    // Recreate the table if it exists, if it's in the list, and if we aren't 
+    // skipping existing ones
+    if ($table_exists and is_object($custom_table) and !$skip_if_exists) {
+      $chado->schema()->dropTable($table);
+      $chado->schema()->createTable($table, $schema);
+      // TODO: the old code unset "referring_tables" from $schema,
+      //       is this necessary?
```
Comment (laceysanderson) [inline tripal_chado/src/api/tripal_chado.custom_tables.api.inc]: We'll want to look into why that was done in the past to ensure its not necessary. Is there a reason not to do it anyway? Does it cause problems or a performance hit?

File: tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        'type_name' => array (
+          'type' => 'varchar',
+          'length' => '1024',
+          'not null' => true,
+        ),
+      ),
+    );
+
+    // IMPORTANT: We decided to cancel testing these features
+    // due to the way in which Drupal 8/9 handles Postgres 'schemas' within a db.
```
Comment (laceysanderson) [inline tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php:71]: Is this still the case or was this comment made before Tripal DBX was merged in? If the second is true then we'll want to mention that these tests should be revived/completed in the extra issue.

File: tripal_chado/tripal_chado.views.inc [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+/**
+ * Describe various Tripal Core systems to Views
+ *
+ * @ingroup tripal
+ */
+function tripal_chado_views_data() {
+  $data = [];
+  // Job Management System.
+  // tripal_chado_views_data_jobs($data);
```
Comment (laceysanderson) [inline tripal_chado/tripal_chado.views.inc]: Why is this removed?

**Review by laceysanderson — DISMISSED** [review]
I did a quick code review as well to make sure this code fits well into the rest of the t4d8 system and nothing jumped out at me 🙂 👍 

I did notice a couple of todo's and comments in the code that look like you have some things to come back to? I'll comment on them specifically but I think they can be addressed in a new PR. Please create an issue with these mentioned so we don't forget about them and expand a bit on your question/motivation for each :-)

PS. I'm going to mark this as approved since this PR looks pretty sound and Stephen already did manual testing. For my feedback, you should open a new issue summarizing things that still need to be looked into or tested -this way you (or someone else) can make an additional PR later cleaning these last things up and adding testing :-)

File: tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php (hunk 6 above)
Comment (spficklin) [reply tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php:71]: Any tests will fail until issue #217 is resolved, unfortunately.

Comment (laceysanderson) [reply tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php:71]: True. Well still good to add to the issue to come back to testing and link the above issue :-)

Comment (spficklin) [reply tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php:71]: Agreed

File: tripal_chado/src/Form/ChadoCustomTableForm.php (hunk 1 above)
Comment (spficklin) [reply tripal_chado/src/Form/ChadoCustomTableForm.php]: I'm jumping in to respond to these. Thanks @laceysanderson for the checks.  Yes, this can be removed. I'll push a change to remove them.

File: tripal_chado/src/Form/ChadoCustomTableForm.php (hunk 2 above)
Comment (spficklin) [reply tripal_chado/src/Form/ChadoCustomTableForm.php]: I believe this was removed because it was a quick check using old PHP style arrays.  Since we now allow the new style this shouldn't be needed. I will take out the commented line and push.

File: tripal_chado/src/api/ChadoSchema.php (hunk 3 above)
Comment (spficklin) [reply tripal_chado/src/api/ChadoSchema.php]: I think it should return NULL. I adjusted it and will push the change.

File: tripal_chado/src/api/ChadoSchema.php (hunk 4 above)
Comment (spficklin) [reply tripal_chado/src/api/ChadoSchema.php]: I think this if statement is old and should come out. The custom table gets selected above at the beginning of the function, so there's no need for this.  I also fixed this function to follow Drupal coding standards too.

File: tripal_chado/src/api/tripal_chado.custom_tables.api.inc (hunk 5 above)
Comment (spficklin) [reply tripal_chado/src/api/tripal_chado.custom_tables.api.inc]: Removing the referring tables key is no longer needed. It seems Drupal just ignores it now. I tested including it when adding a table and it worked just fine. Drupal ignored it.  I've removed the comment and I'll push the update.

File: tripal_chado/tripal_chado.views.inc (hunk 7 above)
Comment (spficklin) [reply tripal_chado/tripal_chado.views.inc]: Hmmm.... that code should not have been there. It's a duplicate of the code in the `tripal` module for creating the Jobs View.  I suspect it may have been there as a reference for creating the Custom Tables view and was not later removed.  I removed it and will push.

File: tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php (hunk 6 above)
Comment (spficklin) [reply tripal_chado/tests/src/Functional/api/ChadoCustomTablesAPITest.php:71]: Issue #223 added.

**Review by laceysanderson — APPROVED** [review]
The new changes all look good 👍 This is ready to merge 🎉
